### ISSUE TITLE
Server-side replication of ~3 GB files retries endlessly over a slower network connection

### ISSUE THREAD

**Issue by sebastiandanconia** [minio/minio#16714]
When server-side replication of buckets is configured and there are objects to be sent to the destination, "larger" objects never finish transferring and instead are retried endlessly. (For the purposes of this bug, "larger" means somewhere between 1 and 4 GB. I was able to reproduce the problem with bandwidth (limited using the `tc` utility) of 10 mbit--50 mbit.)

For test purposes, `Replication Workers` `Replication Failed Workers` in the 'API' section of 'Settings' were each set to 1.

#### Log messages (repeated in a loop, every 10--12 minutes):
```API: SYSTEM()
Time: 04:33:18 UTC 02/25/2023
DeploymentID: ...
Error: Unable to replicate for object testing-site/debian-live-11.4.0-amd64-cinnamon+nonfree.iso(517342eb-eeb5-4299-9dec-447ef53d3eee): Put "http://garrett0.lan:9000/testing/debian-live-11.4.0-amd64-cinnamon%2Bnonfree.iso?versionId=517342eb-eeb5-4299-9dec-447ef53d3eee": write tcp 172.17.0.2:58300->10.NN.NN.NN:9000: use of closed network connection (*errors.errorString)
       3: internal/logger/logger.go:258:logger.LogIf()
       2: cmd/bucket-replication.go:1377:cmd.ReplicateObjectInfo.replicateAll()
       1: cmd/bucket-replication.go:969:cmd.replicateObject.func2()
```
```API: SYSTEM()
Time: 04:43:40 UTC 02/25/2023
DeploymentID: ...
Error: Unable to replicate for object testing-site/debian-live-11.4.0-amd64-cinnamon+nonfree.iso(517342eb-eeb5-4299-9dec-447ef53d3eee): Put "http://garrett0.lan:9000/testing/debian-live-11.4.0-amd64-cinnamon%2Bnonfree.iso?versionId=517342eb-eeb5-4299-9dec-447ef53d3eee": write tcp 172.17.0.2:60684->10.NN.NN.NN:9000: use of closed network connection (*errors.errorString)
       3: internal/logger/logger.go:258:logger.LogIf()
       2: cmd/bucket-replication.go:1377:cmd.ReplicateObjectInfo.replicateAll()
       1: cmd/bucket-replication.go:969:cmd.replicateObject.func2()
```
#### Disk and Network I/O
The network connection from the source to the destination cluster is roughly saturated, as measured by `iftop`.
During one of my tests at 10 mbit, I watched how the file sizes on the destination drive (Erasure backend) changed over time using:
```du -h --summarize /mnt/DISK0/minio/.minio.sys/tmp /mnt/DISK0/minio/.minio.sys/multipart```

`.minio.sys/tmp` would grow up to about 145 GB, then roll over to zero as a new `part.NNN` file was added to `.minio.sys/multipart`. Then eventually, when `.minio.sys/multipart` was 19--50 GB in size, both locations would go down close to zero size, and the data was apparently deleted, since I couldn't account for anywhere it was taking up space on disk using `du`.
Each `part.NNN` in `.minio.sys/multipart/<hash>/<UUID>/<UUID>` is 145 MB.

## Expected Behavior

After configuring server-side bucket replication and adding objects to the source bucket, disk usage in the subdirectories of the destination cluster/drive responsible for receiving incoming objects from server-side replication should monotonically increase until an entire file is transferred and is visible in the Object Browser of the Web UI.

## Current Behavior

Network bandwidth from the source cluster to the destination cluster is nearly saturated, the sizes of `.minio.sys/tmp` and `.minio.sys/multipart` keep changing and getting reset back to low numbers, and log messages appear every 10--15 minutes, however the objects are never transferred successfully and never show up in the destination bucket as visible from the Web UI.

## Possible Solution

Perhaps the log error messages,
`Error: Unable to replicate for object ...` and
`use of closed network connection (*errors.errorString)` cited above hint at a root cause?

## Steps to Reproduce (for bugs)

1. Run Minio in docker on the source and destination servers
2. Install `iftop` on your destination server and `iproute2` on your source server, if necessary.
3. Set up credentials. Source must have credentials to replicate the bucket in question to the destination.
4. In the Settings --> API screen of the source system, set Replication Workers and Replication Failed Workers each to 1.
5. Create buckets on the source and destination systems, (e.g. "testing").
6. Set up `mc` to control the source system (alias/credentials). Run `mc replicate add <source-server-alias>/testing --remote-bucket 'http://<access-key>:<secret-key>@<destination-server>:9000/testing' --replicate 'existing-objects'`
7. Verify the replication is configured correct by uploading a < 1MB file to the source bucket and verifying this file/object is replicated to the source bucket.
8. On the source system run `tc qdisc add dev <iface-name> root handle 1: tbf rate 20mbit latency 30ms burst 8192
9. The test files I used are available at [https://cdimage.debian.org/cdimage/unofficial/non-free/cd-including-firmware/current-live/amd64/iso-hybrid/](url): `debian-live-11.4.0-amd64-cinnamon+nonfree.iso` and `debian-live-11.4.0-amd64-gnome+nonfree.iso`. Upload a test file of 2--4 GB to your source bucket using web UI.
10. Run `iftop -i <iface-name>` on your destination server to verify that data is being received (18 mbps or more).
11. On the source server, do `docker logs --follow minio` to watch the log messages.
12. Wait 15--20 minutes for one or both of the log messages cited above to appear.
13. Wait an hour or so if desired, to convince yourself the file should have transferred by now. The network bandwidth usage will continue, as evidenced by `iftop`.

## Context

I want to set up a server for offsite replication (personal and home office data over residential internet connections).

## Regression

No

## Your Environment

* Version used (`minio --version`): RELEASE.2023-02-22T18-23-45Z (go1.19.6 linux/amd64)
* Server setup and configuration: Ubuntu 22.04.1 LTS (GNU/Linux 5.15.0-60-generic x86_64)
* Both source and destination have Erasure backend. Source has one disk; destination has two servers with three disks each.
* Docker version 23.0.1, build a5ee5b1
* Operating System and version (`uname -a`): Linux hdd-utility 5.15.0-60-generic #66-Ubuntu SMP Fri Jan 20 14:29:49 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

**Comment by harshavardhana** [comment]
@sebastiandanconia will check.

**Comment by sebastiandanconia** [comment]
I left something out of the log messages I quoted in my original post. I meant to post two different messages instead of two of the same. Here's an example of the one I missed:
```API: SYSTEM()
Time: 05:32:36 UTC 02/25/2023
DeploymentID: ...
Error: failed to get lock for object: debian-live-11.4.0-amd64-cinnamon+nonfree.iso bucket:testing-site arn: (*errors.errorString)
       3: internal/logger/logger.go:258:logger.LogIf()
       2: cmd/bucket-replication.go:941:cmd.replicateObject()
       1: cmd/bucket-replication.go:1630:cmd.(*ReplicationPool).AddMRFWorker()
```
The two types of log message, `Unable to replicate for object` and `use of closed network connection (*errors.errorString)`, are interspersed.

**Comment by harshavardhana** [comment]
> 

This is normal @sebastiandanconia  because of the previous attempt. There are two concurrent replication in progress.

Should be fairly simply to understand and reproduce.

**Comment by jcrofts-oci** [comment]
Another thing I missed in my proofreading:
`tc qdisc add dev root handle 1: tbf rate 20mbit latency 30ms burst 8192`
should be:
`tc qdisc add dev <iface-name> root handle 1: tbf rate 20mbit latency 30ms burst 8192`

With `iftop -i <iface-name>`, we're looking for roughly 18--20 Mbps, which shows that the above traffic shaping command is working as intended.

**Comment by sebastiandanconia** [comment]
Another thing I missed in my proofreading:
`tc qdisc add dev root handle 1: tbf rate 20mbit latency 30ms burst 8192`
should be:
`tc qdisc add dev <iface-name> root handle 1: tbf rate 20mbit latency 30ms burst 8192`.

With `iftop -i <iface-name>`, we're looking for something around 18--20 Mbps to show that the above traffic-shaping command is working as intended.

**Comment by poornas** [comment]
Have you tried `mc cp` on this file from this docker container to your target under the same network constraints ? 

From your description of the issue, replication is attempting multipart uploads that proceed up to a point and abort because of network issues, which would delete the parts written thus far and start a new upload because replication is reattempting the operation. Replication re-queues failures repeatedly until it succeeds, which is to be expected.

Even for setting up bandwidth throttling, we require a 100MBps minimum - if network is that slow, you will be better off trying mc mirror.

**Comment by sebastiandanconia** [comment]
@poornas I've added `mc mirror --watch --preserve` to my testing per your suggestion. However, suppose I want to replicate particular buckets (i.e. not the whole site as in `mc mirror site1-alias/ site2-alias/`), and I want to ensure `mc` transfers only one file at a time, or otherwise want to limit bandwidth. `mc` accepts only one SOURCE/TARGET pair; how could I achieve this using client-side replication? I could kick off multiple instances of `mc`, but they wouldn't all be under the same bandwidth regime (`mc mirror --limit-upload` or "one transfer at a time"). Doesn't this argue for server-side replication (which offers configurable `Replication Workers` and `Replication Failed Workers` knobs) being robust enough to handle this use case?

You're saying the replication daemon/process is reattempting the transfer (when it would be better to let the original transfer complete). Is there something like a hard-coded time limit for each transfer? In essence (according to our hypothesis), the replication daemon/process is deciding **_prematurely_** that a transfer has failed.

**Comment by poornas** [comment]
@sebastiandanconia , was asking for`mc cp` of just this file with the same traffic shaping - if you see issues there, replication can do no better. 
fwiw, you can tinker with setting env var MINIO_CONN_READ_DEADLINE and MINIO_CONN_WRITE_DEADLINE which by default is 10 minutes and see if a longer deadline helps - you will need to restart your instance

**Comment by sebastiandanconia** [comment]
@poornas The test I ran tonight was ```mc mirror --watch --preserve site1-alias/testing site2-alias/testing```, run on the SOURCE computer `site1` (and thus comparably bandwidth-limited as my server-side replication tests were), however outside a Docker container. I haven't configured anything special in terms of Docker networking (i.e. Docker networking was all default settings). Both the Debian install ISOs referenced in the original post transferred without incident. If you feel this is not sufficiently close to the test you requested, please let me know.

It's great to know about `MINIO_CONN_READ_DEADLINE` and `MINIO_CONN_WRITE_DEADLINE`, and I'll be sure to re-test against the next release when it comes out.

**Comment by mymeatball** [comment]
> @sebastiandanconia , was asking for`mc cp` of just this file with the same traffic shaping - if you see issues there, replication can do no better. fwiw, you can tinker with setting env var MINIO_CONN_READ_DEADLINE and MINIO_CONN_WRITE_DEADLINE which by default is 10 minutes and see if a longer deadline helps - you will need to restart your instance

I saw the same error log in my cluster log, and the object failed to replicate after 10 minutes.
I tried setting env var MINIO_CONN_READ_DEADLINE and MINIO_CONN_WRITE_DEADLINE, but it didn't seem to solve the problem.
The steps I tried are as follows:
  1.Setting env var MINIO_CONN_READ_DEADLINE and MINIO_CONN_WRITE_DEADLINE in the file which /etc/profile.Like this:
     export MINIO_CONN_READ_DEADLINE=3600
     export MINIO_CONN_WRITE_DEADLINE=3600
  2.Execute Command：source /etc/profile
  3.Execute Command：systemctl restart minio（All 4 nodes in the cluster perform this operation）

Is there a problem with my steps in setting environment variables? Or is this method useless?

I understand that during the copying process, due to network bandwidth limitations, the file copying took more than 10 minutes but was not completed, which caused the problem to occur.How should I solve this problem besides providing higher bandwidth?

**Comment by mymeatball** [comment]
@poornas Can you help me reslove this problem?

**Comment by daft-panda** [comment]
Would suggest documenting somewhere Minio replication does not work across "slow" WAN links. I've burned though 1.5TB during 5 days transferring 80GB. After no more progress was made I synced using rclone which happily transferred the remaining 30GB of files in 2h30.

**Comment by evandeaubl** [comment]
I've discovered a workaround if you have some level of control over how objects are put into storage. Server-side replication transfers an object in the same way it was uploaded (at least as of the current code), so if an object was originally multipart uploaded, replication will also multipart upload it to the remote with the same chunk size. Once I realized that, I reduced the size threshold for using multipart uploads on the initial PUT (for `rclone`, it's `--s3-upload-cutoff`), and that basically stabilized replication for me over a ~40Mbps link, because those chunk uploads finish well under the 10-minute timeout unlike full file uploads.

Hoping this helps out people who find this issue like I did!
